message/message-factory.rst

```diff
@@ -1,57 +1,140 @@
-
-.. code:: php
-
-    // Create a PSR-7 request
-    $factory = MessageFactoryDiscovery::find();
-    $request = $factory->createRequest('GET', 'http://example.com');
+Usage
+-----
 
+Instantiate the factories in your bootstrap code or use discovery for them.
+Inject the factories into the rest of your code to limit the implementation
+choice to the bootstrapping code::
+
+    // ApiClient.php
+
+    use Http\Message\RequestFactory;
+    use Http\Message\StreamFactory;
+    use Http\Message\UriFactory;
+
+    class ApiClient
+    {
+        /**
+         * @var RequestFactory
+         */
+        private $requestFactory;
+
+        /**
+         * @var StreamFactory
+         */
+        private $streamFactory;
+
+        /**
+         * @var UriFactory
+         */
+        private $uriFactory;
+
+        public function __construct(
+            RequestFactory $requestFactory,
+            StreamFactory $streamFactory,
+            UriFactory $uriFactory
+        ) {
+            $this->requestFactory = $requestFactory;
+            $this->streamFactory = $streamFactory;
+            $this->uriFactory = $uriFactory;
+        }
+
+        public function doStuff()
+        {
+            $request = $this->requestFactory->createRequest('GET', 'http://httplug.io');
+            $stream = $this->streamFactory->createStream('stream content');
+            $uri = $this->UriFactory->createUri('http://httplug.io');
+            ...
+        }
+    }
+
+The bootstrapping code could look like this::
+
+    // bootstrap.php
+    use Http\Message\MessageFactory\DiactorosMessageFactory;
+    use Http\Message\StreamFactory\DiactorosStreamFactory;
+    use Http\Message\UriFactory\DiactorosUriFactory;
+
+    $apiClient = new ApiClient(
+        new DiactorosMessageFactory(),
+        new DiactorosStreamFactory(),
+        new DiactorosUriFactory()
+    );
+
+You could also use :doc:`/discovery` to make the factory arguments optional and
+automatically find an available factory in the client::
+
+    // ApiClient.php
+
+    use Http\Discovery\MessageFactoryDiscovery;
+    use Http\Discovery\StreamFactoryDiscovery;
+    use Http\Discovery\UriFactoryDiscovery;
+    use Http\Message\RequestFactory;
+    use Http\Message\StreamFactory;
+    use Http\Message\UriFactory;
+
+    class ApiClient
+    {
+        public function __construct(
+            RequestFactory $requestFactory = null,
+            StreamFactory $streamFactory = null,
+            UriFactory $uriFactory = null
+        ) {
+            $this->requestFactory = $requestFactory ?: MessageFactoryDiscovery::find(),
+            $this->streamFactory = $streamFactory ?: StreamFactoryDiscovery::find();
+            $this->uriFactory = $uriFactory ?: UriFactoryDiscovery::find();;
+        }
+
+        ...
+    }
+
+.. hint::
+
+    If you create requests only and no responses, use ``RequestFactory`` in the
+    type hint, instead of the ``MessageFactory``. And vice versa if you create
```

Comment:
Already see this in english, you can also use "And conversely" if you want, but i think "vice versa" is better.